Add libvirt podvm image support

supritai opened the pull request:
<!--
If this is a bug fix, make sure your description includes "Fixes: #xxxx", or
"Closes: #xxxx"

Please provide the following information:
-->
**- Description of the problem which is fixed/What is the use case**
Add libvirt image creation job and libvirt image type for image generation.

**- What I did**

**- How to verify it**

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

supritai commented on config/peerpods/podvm/Dockerfile.podvm-builder:
```diff
@@ -29,6 +27,9 @@ RUN if [[ -n "$CERT_RPM" ]] ; then \
 
 RUN git clone ${CAA_SRC} -b ${CAA_REF} /src/cloud-api-adaptor
 
+ADD rhel-guest-image-9.2-20230414.17.s390x.qcow2 /src/cloud-api-adaptor/src/cloud-api-adaptor/podvm/rhel-guest-image-9.2-20230414.17.s390x.qcow2
```
Need to change : download the image while running the job than bundling it along the builder image

supritai55 replied:
I will update the required changes in the next patch

supritai commented on config/peerpods/podvm/libvirt-podvm-image-cm.yaml:
```diff
@@ -0,0 +1,34 @@
+  YQ_CHECKSUM: "sha256:4e6324d08630e7df733894a11830412a43703682d65a76f1fc925aac08268a45"
+  ARCH: "s390x"
+  GO_VERSION: "1.21.9"
+  IMAGE_URL: "rhel-guest-image-9.2-20230414.17.s390x.qcow2"
```
Need to change : download the image while running the job than bundling it along the builder image

supritai55 replied:
I will update the required changes in the next patch

supritai commented on config/peerpods/podvm/libvirt-podvm-image-handler.sh:
```diff
@@ -0,0 +1,220 @@
+    else 
+    echo "Couldn't upload the image." 
+    fi
+    sleep 120s
```
Remove it

supritai55 replied:
done

supritai commented on config/peerpods/podvm/libvirt-podvm-image-handler.sh:
```diff
@@ -0,0 +1,220 @@
+    make image
+    sleep 120s
+    cp -pr /src/cloud-api-adaptor/src/cloud-api-adaptor/podvm/output/*.qcow2 /src/cloud-api-adaptor/src/cloud-api-adaptor/podvm/output/libvirt-podvm.qcow2
+    sleep 120s
```
Remove it

supritai55 replied:
done

supritai commented on config/peerpods/podvm/osc-podvm-create-job.yaml:
```diff
@@ -13,9 +13,11 @@ spec:
       initContainers:
         - name: copy
-          image: registry.redhat.io/openshift-sandboxed-containers/osc-podvm-payload-rhel9:1.7.0
+          image: quay.io/suprit_iti/openshift-sandboxed-containers/osc-podvm-payload-rhel9:latest
```
Change the image source to registry.redhat.io and use it once the changes related to Dockerfile is approved

supritai55 replied:
done

supritai commented on config/peerpods/podvm/osc-podvm-create-job.yaml:
```diff
@@ -29,12 +31,11 @@ spec:
           # Binaries like kubectl, packer and yq are expected to be under /usr/local/bin
           # podvm binaries are expected to be under /payload/podvm-binaries.tar.gz
-          image: registry.redhat.io/openshift-sandboxed-containers/osc-podvm-builder-rhel9:1.7.0
+          image: quay.io/suprit_iti/openshift-sandboxed-containers/osc-podvm-builder-rhel9:testv29
```
Change the image source to registry.redhat.io and use it once the changes related to Dockerfile is approved

supritai55 replied:
I will update the required changes in the next patch

savitrilh commented on config/peerpods/podvm/Dockerfile.podvm-builder:
```diff
@@ -29,6 +27,9 @@ RUN if [[ -n "$CERT_RPM" ]] ; then \
 ADD podvm-builder.sh /podvm-builder.sh
 
-ENTRYPOINT ["/podvm-builder.sh"]
+ENTRYPOINT ["/podvm-builder.sh"]
```
No change here, it is showing as changed. Can you correct this

supritai55 replied:
ok

savitrilh commented on controllers/image_generator.go:
```diff
@@ -200,6 +202,7 @@ func ImageDelete(c client.Client) (int, error) {
 }
 
 func newImageGenerator(client client.Client) (*ImageGenerator, error) {
+	fmt.Println("triggering NewImageGenerator")
```
Remove the printf and add a log

supritai55 replied:
ok

savitrilh commented on controllers/utils.go:
```diff
@@ -168,11 +168,12 @@ func getCloudProviderFromInfra(c client.Client) (string, error) {
+		// For libvirt provider, PlatformStatus.Type is empty string, so returning libvirt provider
+		return "libvirt", nil
 	}
-
```
don't remove newline

supritai55 replied:
ok

savitrilh commented on controllers/utils.go:
```diff
@@ -168,11 +168,12 @@ func getCloudProviderFromInfra(c client.Client) (string, error) {
 	if err != nil {
 		return "", err
 	}
-
```
don't remove newline

supritai55 replied:
ok

savitrilh commented on config/peerpods/podvm/osc-podvm-create-job.yaml:
```diff
@@ -49,11 +50,45 @@ spec:
               mountPath: /payload
+            - mountPath: "/root/.ssh/"
+              name: ssh-key-secret
+              readOnly: true
```
Cann't we do -mountpath as optional as it is not used by other cloud providers

supritai55 replied:
ok

supritai55 commented:
This PR is ready for review. Since I do not have write access, unable to remove the 'draft' tag. Can someone remove this tag please?

bpradipt commented:
> This PR is ready for review. Since I do not have write access, unable to remove the 'draft' tag. Can someone remove this tag please?

Since you are the PR creator, you should be able to change this "Ready".  I believe you raised this PR as a draft pull request.

bpradipt commented:
@supritai @savitrilh can you please rebase the PR..

supritai55 commented:
> @supritai @savitrilh can you please rebase the PR..

Sure @bpradipt

bpradipt commented on config/peerpods/podvm/libvirt-podvm-image-cm.yaml:
```diff
@@ -0,0 +1,34 @@
+  DOWNLOAD_SOURCES: "no"
+  CONFIDENTIAL_COMPUTE_ENABLED: "no"
+
+  # Livirt specific
```
s/Livirt/Libvirt

supritai replied:
Okay, will change

bpradipt commented on config/peerpods/podvm/libvirt-podvm-image-cm.yaml:
```diff
@@ -0,0 +1,34 @@
+  YQ_VERSION : "v4.35.1"
+  YQ_CHECKSUM: "sha256:4e6324d08630e7df733894a11830412a43703682d65a76f1fc925aac08268a45"
+  ARCH: "s390x"
+  GO_VERSION: "1.21.9"
```
Is GO_VERSION needed here ?

supritai replied:
Yes, As using this variable installing GO as part of 'config/peerpods/podvm/libvirt-podvm-image-handler.sh:128'

snir911 replied:
IIUC yq and go are used only in order build the tools right? (i.e. not peerpods/coco components) in that  case, i think it would be ok to use fixed versions in script and don't allow the user to set these.

supritai replied:
GO, 'Development Tools'(gcc) - used for building umoci binary
yq, genisoimage and cloud-utils - for packer 
Agree, these can be fixed version than user setting up.

bpradipt commented on config/peerpods/podvm/libvirt-podvm-image-handler.sh:
```diff
@@ -0,0 +1,220 @@
+    cd "${CAA_SRC_DIR}"/src/cloud-api-adaptor/podvm || \
+	    error_exit "Failed to change directory to ${CAA_SRC_DIR}/src/cloud-api-adaptor/podvm"
+    make image
+    sleep 120s
```
What's the reason for sleep here ?

supritai replied:
removed it

bpradipt commented on config/peerpods/podvm/libvirt-podvm-image-handler.sh:
```diff
@@ -0,0 +1,220 @@
+    sleep 120s
+    cp -pr /src/cloud-api-adaptor/src/cloud-api-adaptor/podvm/output/*.qcow2 /src/cloud-api-adaptor/src/cloud-api-adaptor/podvm/output/libvirt-podvm.qcow2
+    sleep 120s
+    #mkdir -p /src/cloud-api-adaptor/src/cloud-api-adaptor/podvm/output
```
Leftover ?

supritai replied:
yes, removed

snir911 commented on config/peerpods/podvm/libvirt-podvm-image-cm.yaml:
```diff
@@ -0,0 +1,34 @@
+  ACTIVATION_KEY: ""
+  YQ_VERSION : "v4.35.1"
+  YQ_CHECKSUM: "sha256:4e6324d08630e7df733894a11830412a43703682d65a76f1fc925aac08268a45"
+  ARCH: "s390x"
```
Have you tested with x86 also by any chance? i don't see a reason why it wouldn't work, it will allow easy testing.

supritai replied:
Sorry, I have not tested on x86.

snir911 commented on config/peerpods/podvm/libvirt-podvm-image-handler.sh:
```diff
@@ -0,0 +1,220 @@
+        install_rpm_packages
+
+        # Install required binary packages
+        install_libvirt_binary_packages
```
This function is not defined

supritai replied:
This is not required, removed it.

snir911 commented on config/peerpods/podvm/libvirt-podvm-image-handler.sh:
```diff
@@ -0,0 +1,220 @@
+
+    export PODVM_DISTRO=${PODVM_DISTRO}
+    export CLOUD_PROVIDER=${CLOUD_PROVIDER}
+    export ARCH=${ARCH}
```
If exporting for this 3 vars is required I'd guess it's required also for the others no? (alternatively maybe exporting is not required?)

supritai replied:
These are used as environment variables for https://github.com/confidential-containers/cloud-api-adaptor/blob/db56a2d3b30ff7e42cc100e6616ef049cdcc03b8/src/cloud-api-adaptor/podvm/Dockerfile.podvm.rhel#L23

snir911 replied:
right, but i don't think it's required to export them as they are already defined

supritai replied:
Right, will remove them

snir911 commented on config/peerpods/podvm/libvirt-podvm-image-handler.sh:
```diff
@@ -0,0 +1,220 @@
+    [[ -z "${CLOUD_PROVIDER}" || "${CLOUD_PROVIDER}" != "libvirt" ]] && error_exit "CLOUD_PROVIDER is empty or not set to libvirt"
+
+    [[ -z "${ORG_ID}" ]] && error_exit "ORG_ID is not set"
+    [[ -z "${ACTIVATION_KEY}" ]] && error_exit "ACTIVATION_KEY is not set"
```
are these always mandatory ?  i mean. is it possible that image is already registered in certain cases? or that the installing the pkgs doesn't really requires registering?

supritai replied:
It looks mandatory for the builder image which I have been using. If the builder image registered, then no need of it. https://github.com/openshift/sandboxed-containers-operator/blob/a77c2e93f2e46d8e0b568f92a3ff51be938eab50/config/peerpods/podvm/Dockerfile.podvm-builder#L1

snir911 replied:
I'd check again if this is still mandatory as IIUC some of the pkgs installation were dropped

supritai replied:
If the host is a subscribed, no need for subscription of the container.
If I remove the subscription and test :
While running the job, neither worker node nor container would be registered and package installation fails.
```
_This system is not registered with an entitlement server. You can use subscription-manager to register.

Last metadata expiration check: 17:37:29 ago on Tue Jul 16 12:23:50 2024.
Module or Group 'Development Tools' is not available.
Error: Nothing to do.
++ dnf install -y genisoimage qemu-kvm libvirt-client
Updating Subscription Management repositories.
Unable to read consumer identity

This system is not registered with an entitlement server. You can use subscription-manager to register.

Last metadata expiration check: 17:37:30 ago on Tue Jul 16 12:23:50 2024.
No match for argument: genisoimage
No match for argument: qemu-kvm
No match for argument: libvirt-client
```

snir911 commented on config/peerpods/podvm/libvirt-podvm-image-handler.sh:
```diff
@@ -0,0 +1,220 @@
+    # If any error occurs, exit the script with an error message
+
+    # LIBVIRT_ID shouldn't be empty
+    [[ -z "${LIBVIRT_ID}" ]] && error_exit "LIBVIRT_ID is empty"
```
where LIBVIRT_ID is coming from? how it's being set?

supritai replied:
LIBVIRT_POOL_UUID should be supplied as env variable

snir911 commented on config/peerpods/podvm/libvirt-podvm-image-handler.sh:
```diff
@@ -0,0 +1,220 @@
+    echo "-R Recreate podvm-images configMap"
+}
+
+function install_packages(){
```
all of these are actually needed? are you building binaries from source or being based on payload container? I'd expect mostly the libvirt pkgs & packer to be needed

supritai replied:
Binaries are taken from the payload. Yes, mostly libvirt related packages are need to be installed.

snir911 replied:
please remove all of those which aren't required (go, rust etc..)

supritai replied:
go is required for building the some of the packages:
Ref: https://github.com/confidential-containers/cloud-api-adaptor/blob/0480804f8f7d06202a5a18fef3a966ec6e4e59c7/src/cloud-api-adaptor/podvm/Makefile.inc#L125

supritai replied:
rust not needed, removed

snir911 replied:
> go is required for building the some of the packages:
Ref: https://github.com/confidential-containers/cloud-api-adaptor/blob/0480804f8f7d06202a5a18fef3a966ec6e4e59c7/src/cloud-api-adaptor/podvm/Makefile.inc#L125

The forwarder should be included in the payload image, hence, it should not be built as part of creating the image process, if it is, something might have been missed in the process of building the payload image

supritai replied:
My bad, GO is not required there. It is required for umoci ( as need to be installed from source due to unavailability of s390x package)

snir911 commented on config/peerpods/podvm/libvirt-podvm-image-cm.yaml:
```diff
@@ -0,0 +1,34 @@
+  PROTOC_VERSION: "26.1"
+  PODVM_TAG: "v1.0"
+  PACKER_VERSION: "v1.9.4"
+  RUST_VERSION: "1.72.0"
```
See my comment on the installed pkgs, i suspect all of the versions are not needed

snir911 commented on config/peerpods/podvm/libvirt-podvm-image-handler.sh:
```diff
@@ -0,0 +1,220 @@
+
+function install_packages(){
+    if [[ -n "${ACTIVATION_KEY}" && -n "${ORG_ID}" ]]; then \
+    subscription-manager register --org=${ORG_ID} --activationkey=${ACTIVATION_KEY}; \
```
nit, I think the style is commonly to indent the lines within conditionals

supritai replied:
sure, will indent wherever required

snir911 commented on config/peerpods/podvm/libvirt-podvm-image-handler.sh:
```diff
@@ -0,0 +1,220 @@
+    if [ $? -eq 0 ]; then
+    echo "Uploaded the image successfully" 
+    LIBVIRT_POOL_UUID=$(virsh -d 0 -c $LIBVIRT_URI pool-list --name --uuid | grep $LIBVIRT_POOL | awk '{print $1}')
+    echo "Updating peer-pods-cm configmap with IMAGE_ID=${LIBVIRT_POOL_UUID}" 
```
where's the actual update of the cm?

supritai replied:
I will add function to add the same.

snir911 commented on config/peerpods/podvm/Dockerfile.podvm-builder:
```diff
@@ -32,4 +32,4 @@ ADD rhel-guest-image-9.2-20230414.17.s390x.qcow2 /src/cloud-api-adaptor/src/clou
 
-ENTRYPOINT ["/podvm-builder.sh"]
\ No newline at end of file
+ENTRYPOINT ["/podvm-builder.sh"]
```
nit, if possible pls squash the addressed comments and resolve them in github

supritai replied:
I will do

snir911 commented:
Thanks! I like the idea, added few comments

bpradipt commented:
/ok-to-test

supritai commented on config/peerpods/podvm/libvirt-podvm-image-handler.sh:
```diff
@@ -0,0 +1,273 @@
+    fi
+
+    # Add the libvirt pool uuid as annotation to peer-pods-cm configmap
+    kubectl patch configmap peer-pods-cm -n openshift-sandboxed-containers-operator \
```
@snir911 awa/azure are adding as the annotation : https://github.com/openshift/sandboxed-containers-operator/blob/b13fc7111e920795df3a6f7a9e6e734cfa2539a9/config/peerpods/podvm/aws-podvm-image-handler.sh#L263 
But getImageID method looks for Data in configmap. Not sure how it is works.
https://github.com/openshift/sandboxed-containers-operator/blob/b13fc7111e920795df3a6f7a9e6e734cfa2539a9/controllers/image_generator.go#L556
I have patched as Data field in configmap.

snir911 replied:
Right, in Azure and AWS the IMAGE_ID in the CM is used in order to identify the created podvm image, however, in libvirt, the equivalent is the volume name, which by default is "podvm-base.qcow2", unless LIBVIRT_VOL_NAME is set.
Actually i think you should drop LIBVIRT_POOL_UUID completely in favor of LIBVIRT_VOL_NAME, as the uuid is not consumed by CAA. 
cc @bpradipt for advise

bpradipt replied:
I agree.. if LIBVIRT_POOL_UUID is not used, then there is no point in adding.
If this will be required later to support per pod VM image, it can be added later

savitrilh replied:
We need one Libvirt related ID( CMimageIDKey)  LIBVIRT_POOL_UUID or LIBVIRT_VOL_NAME. This check is  required whether to  skip or perform the image configuration operation. 

We thought to use LIBVIRT_POOL_UUID_<Vol NAME> as a CMimageIDKey to differentiate between different Libvirt images. We planned to do corresponding changes in CAA also. 
This decided this way because two different LibVirt Pool can have same Volume names. This differentiation will be useful to support multiple PodVM images.

Now we will set CMimageIDKey = LIBVIRT_VOL_NAME for libVirt provider. In future, we will change as we said above.

Is it okay?

bpradipt replied:
yeah sounds good. Let's use LIBVIRT_VOL_NAME for now

bpradipt commented on config/peerpods/peerpodssecret.yaml:
```diff
@@ -9,3 +9,4 @@ stringData:
   #LIBVIRT_URI: "qemu+ssh://root@192.168.122.1/system?no_verify=1"
   #LIBVIRT_NET: "default
   #LIBVIRT_POOL: "default"
+  #REDHAT_OFFLINE_TOKEN: ""
```
What is this token for? Please add a comment describe the use of this token.

supritai replied:
This is to download base rhel image, I will add the comment

bpradipt commented on config/peerpods/podvm/libvirt-podvm-image-cm.yaml:
```diff
@@ -0,0 +1,31 @@
+  ACTIVATION_KEY: ""
+  YQ_VERSION : "v4.35.1"
+  YQ_CHECKSUM: "sha256:4e6324d08630e7df733894a11830412a43703682d65a76f1fc925aac08268a45"
+  ARCH: "s390x"
```
Why to set ARCH default to `s390x` here? Also is ARCH needed?  
Can't it be derived in the handler itself by running `uname -m` ?

supritai replied:
okay, removed ARCH var

bpradipt commented on config/peerpods/podvm/libvirt-podvm-image-cm.yaml:
```diff
@@ -0,0 +1,31 @@
+  PROTOC_VERSION: "26.1"
+  PODVM_TAG: "v1.0"
+  PACKER_VERSION: "v1.9.4"
+  OS_VERSION: "9.2"
```
I don't think PROTOC and PACKER version is needed here. PROTOC will be used for binary builds and will be part of payload. PACKER version is not configurable as we need to stick to 1.9.4

bpradipt replied:
Also is OS_VERSION the base OS version? if yes, then may be the var can be called as BASE_OS_VERSION or similar ?

snir911 replied:
Is PROTOC_VERSION needed? AFAIU we never build binaries at runtime no?

bpradipt replied:
Right

supritai replied:
Yes not needed, removed

bpradipt commented on config/peerpods/podvm/libvirt-podvm-image-handler.sh:
```diff
@@ -0,0 +1,301 @@
+    dnf install -y yum-utils gnupg git --allowerasing curl pkg-config clang perl libseccomp-devel gpgme-devel \
+    device-mapper-devel qemu-kvm unzip wget libassuan-devel genisoimage cloud-utils-growpart cloud-init;
+
+    curl -L -o /usr/local/bin/yq https://github.com/mikefarah/yq/releases/download/${YQ_VERSION}/yq_linux_${ARCH/x86_64/amd64} \
```
I would recommend updating the install_binary_packages() in lib.sh to support multi-arch.
For binaries which you need to build may be have a separate method in lib.sh and use it in the handler.

bpradipt replied:
If some binaries can be built downstream and made part of the builder image it will be good. Otherwise the process will be so time consuming. Wdyt @snir911 ?

snir911 replied:
I agree, however, @supritai said binaries are taken from payload image, hence, I'm not sure what is left to be built here, IIUC most of the preparations in this file can be dropped (these are taken from the upstream Dockerfile.podvm_builder*)

supritai replied:
Included only needed packages.

gkurz commented:
General remarks :
- I see multiple merge commits in this PR. Please reorganize this PR to be a list of regular commits on top of current devel.
- Please amend all commits to have proper description and Signed-off-by tags

gkurz commented:
@supritai please do proper rebases instead of merging upstream devel again and again in your repo. We don't want this complexity in our `git log`. A PR should be a list of consecutive regular non-merge commits that apply to the target branch.

Each commit must have a description that helps understanding the change and have proper DCO (Signed-off-by).

This is hard requirement for this PR to get merged in addition to the requests from reviewers.

supritai commented:
> @supritai please do proper rebases instead of merging upstream devel again and again in your repo. We don't want this complexity in our `git log`. A PR should be a list of consecutive regular non-merge commits that apply to the target branch.
> 
> Each commit must have a description that helps understanding the change and have proper DCO (Signed-off-by).
> 
> This is hard requirement for this PR to get merged in addition to the requests from reviewers.

@gkurz Sure. I will do that. I am waiting for review to reach final stage so that I can squash all the commits.

bpradipt commented on controllers/image_generator.go:
```diff
@@ -295,6 +301,11 @@ func (r *ImageGenerator) createJobFromFile(jobFileName string) (*batchv1.Job, er
 			})
+		} else if r.provider == LibvirtProvider {
+			job.Spec.Template.Spec.Containers[0].Env = append(job.Spec.Template.Spec.Containers[0].Env, corev1.EnvVar{
+				Name:  "LIBVIRT_VOL_NAME",
```
nit 's/LIBVIRT_VOL_NAME/peerpodsLibvirtImageKey`

snir911 commented on config/peerpods/podvm/libvirt-podvm-image-handler.sh:
```diff
@@ -0,0 +1,289 @@
+    export IMAGE_URL=$CAA_SRC_DIR/podvm/rhel-$BASE_OS_VERSION-s390x-kvm.qcow2
+    export IMAGE_CHECKSUM=$(sha256sum ${IMAGE_URL} | awk '{ print $1 }')
+
+}
```
nice! maybe use $ARCH instead of s390x for future muliarch support

supritai replied:
okay

snir911 commented on config/peerpods/peerpodssecret.yaml:
```diff
@@ -9,3 +9,4 @@ stringData:
   #LIBVIRT_URI: "qemu+ssh://root@192.168.122.1/system?no_verify=1"
   #LIBVIRT_NET: "default
   #LIBVIRT_POOL: "default"
+  #REDHAT_OFFLINE_TOKEN: "" #Required to download rhel base image
```
maybe mention also how to get it (https://access.redhat.com/management/api IIUC)

snir911 replied:
Will it support getting the image from container payload? if not, consider validating REDHAT_OFFLINE_TOKEN is set (IIUC it will be mandatory)

supritai replied:
No, I will add condition to check the variable

snir911 commented on config/peerpods/podvm/libvirt-podvm-image-handler.sh:
```diff
@@ -0,0 +1,289 @@
+    # Install openshift-client
+    mkdir -p /tmp/openshift-client
+    curl https://mirror.openshift.com/pub/openshift-v4/s390x/clients/ocp/4.2.13-s390x/openshift-client-linux-4.2.13-s390x.tar.gz -o /tmp/openshift-client/openshift-client-linux-4.2.13-s390x.tar.gz
+    tar -xvf /tmp/openshift-client/openshift-client-linux-4.2.13-s390x.tar.gz -C /tmp/openshift-client
```
nit, no pkg for openshfit client?

supritai replied:
It is there. I have included in install_binary_packages() in lib.sh. I will remove this redundant code.

snir911 reviewed (commented):
thanks!

snir911 commented on config/peerpods/podvm/libvirt-podvm-image-cm.yaml:
```diff
@@ -0,0 +1,25 @@
+  # Libvirt specific
+  ORG_ID: ""
+  ACTIVATION_KEY: ""
+  PODVM_TAG: "v1.0"
```
redundant?

supritai replied:
yes, removed

bpradipt commented on config/peerpods/podvm/lib.sh:
```diff
@@ -72,13 +73,22 @@ function install_rpm_packages() {
+    if [ "${ARCH}" == "s390x" ]; then
+        # Define the required binary packages
+        REQUIRED_BINARY_PACKAGES=(
+            "kubectl=https://mirror.openshift.com/pub/openshift-v4/s390x/clients/ocp/4.2.13-s390x/openshift-client-linux-4.2.13-s390x.tar.gz"
```
You can use a similar url like x86_64 with just the arch being different - https://mirror.openshift.com/pub/openshift-v4/s390x/clients/ocp/4.14.9/openshift-client-linux.tar.gz

bpradipt replied:
```suggestion
            "kubectl=https://mirror.openshift.com/pub/openshift-v4/s390x/clients/ocp/4.14.9/openshift-client-linux.tar.gz"
```

supritai replied:
okay

bpradipt commented on config/peerpods/podvm/lib.sh:
```diff
@@ -110,9 +120,9 @@ install_binary_packages() {
+            # yq extracted file name is yq_linux_${ARCH}. Rename it
             [[ "${package_name}" == "yq" ]] &&
-                mv "${install_dir}/yq_linux_amd64" "${install_dir}/yq"
+                mv "${install_dir}/yq_linux_${ARCH}" "${install_dir}/yq"
```
This will fail for x86_64 since ARCH will be `x86_64` whereas the expected filename has `amd64`

supritai replied:
changed

bpradipt commented on config/peerpods/podvm/podvm-builder.sh:
```diff
@@ -17,6 +17,13 @@ function install_aws_deps() {
+function install_libvirt_deps() {
+  echo "Installing libvirt deps"
+  # Install the required packages
+  . /scripts/libvirt-podvm-image-handler.sh -- install_binaries
```
```suggestion
/scripts/libvirt-podvm-image-handler.sh -- install_binaries
```

supritai replied:
okay

bpradipt commented on config/peerpods/podvm/osc-podvm-delete-job.yaml:
```diff
@@ -27,6 +27,8 @@ spec:
               value: "" # Set this to the aws ami id to delete
             - name: IMAGE_ID
               value: "" # Set this to the azure image id to delete
+            - name: LIBVIRT_ID
```
I think you can use IMAGE_ID itself. LIBVIRT_ID sounds incorrect.

supritai replied:
okay

bpradipt commented on config/peerpods/podvm/podvm-builder.sh:
```diff
@@ -181,9 +199,42 @@ function delete_podvm_image() {
+    ;;
+    libvirt)
+    # If LIBVIRT_ID is not set, then exit
+    if [ -z "${LIBVIRT_ID}" ]; then
```
you can use IMAGE_ID, which will be more apt

supritai replied:
okay

bpradipt commented on config/peerpods/podvm/podvm-builder.sh:
```diff
@@ -181,9 +199,42 @@ function delete_podvm_image() {
+
+    # If LIBVIRT_IMAGE_ID is not set, then exit
+    if [ -z "${LIBVIRT_IMAGE_ID}" ]; then
+      echo "PODVM_AMI_ID is not set in peer-pods-cm. Skipping the deletion of AWS AMI"
```
```suggestion
echo "LIBVIRT_IMAGE_ID is not set in peer-pods-cm. Skipping the deletion of the libvirt image"
```

supritai replied:
corrected

bpradipt commented on config/peerpods/podvm/podvm-builder.sh:
```diff
@@ -181,9 +199,42 @@ function delete_podvm_image() {
+    echo "Deleting Libvirt image id"
+    /scripts/libvirt-podvm-image-handler.sh -C
+
+    # Update the peer-pods-cm configmap and remove the PODVM_AMI_ID value
```
s/PODVM_AMI_ID/LIBVIRT_IMAGE_ID

supritai replied:
sorry, missed it

bpradipt commented on config/peerpods/podvm/podvm-builder.sh:
```diff
@@ -181,9 +199,42 @@ function delete_podvm_image() {
+      exit 1
+    fi
+
+    # check if the LIBVIRT_IMAGE_ID value in peer-pods-cm is same as the input LIBVIRT_IMAGE_ID
```
```suggestion
     # check if the LIBVIRT_IMAGE_ID value in peer-pods-cm is same as the input IMAGE_ID
```

supritai replied:
ok

bpradipt commented on config/peerpods/podvm/libvirt-podvm-image-cm.yaml:
```diff
@@ -0,0 +1,25 @@
+  name: libvirt-podvm-image-cm
+  namespace: openshift-sandboxed-containers-operator
+data:
+  CLOUD_PROVIDER: "libvirt"
```
not needed

supritai replied:
okay

bpradipt commented on config/peerpods/podvm/libvirt-podvm-image-handler.sh:
```diff
@@ -0,0 +1,280 @@
+    [[ -z "${CAA_SRC}" ]] && error_exit "CAA_SRC is empty"
+    [[ -z "${CAA_REF}" ]] && error_exit "CAA_REF is empty"
+
+    [[ -z "${REDHAT_OFFLINE_TOKEN}" ]] && error_exit "Redhat token is not set"
```
REDHAT_OFFLINE_TOKEN is not there in `libvirt-podvm-image-cm`

supritai replied:
REDHAT_OFFLINE_TOKEN it is in peerpod secret.

snir911 replied:
I assume this is sensitive value, wouldn't be better to keep in the  Secret?

bpradipt commented on config/peerpods/podvm/libvirt-podvm-image-handler.sh:
```diff
@@ -0,0 +1,280 @@
+    install_binary_packages
+
+    #Install other libvirt specific binaries packages
+    export ARCH=$(uname -m)
```
please split command o/p assignment and export into separate statements
```
ARCH=$(uname -m)
export ARCH
```

supritai replied:
okay

bpradipt commented on config/peerpods/podvm/libvirt-podvm-image-handler.sh:
```diff
@@ -0,0 +1,280 @@
+    dnf install -y gcc genisoimage qemu-kvm libvirt-client
+
+    GO_VERSION="1.21.9"
+    curl https://dl.google.com/go/go${GO_VERSION}.linux-${ARCH/x86_64/amd64}.tar.gz -o go${GO_VERSION}.linux-${ARCH/x86_64/amd64}.tar.gz && \
```
please quote all shell vars eg `"${ORG_ID}"`

supritai replied:
okay

bpradipt commented on config/peerpods/podvm/libvirt-podvm-image-handler.sh:
```diff
@@ -0,0 +1,280 @@
+        make bin && cp bin/packer /usr/local/bin/
+        cd $OLDPWD
+    else
+        yum-config-manager --add-repo https://rpm.releases.hashicorp.com/RHEL/hashicorp.repo && \
```
packer is already being installed for x86_64 as part of install_binary_packages. So this is not needed here

supritai replied:
But we need packer for s390 systems. Am I missing something?

bpradipt replied:
You are already building packer here for s390x. When this code runs on x86_64, then the original install_binary_packages (which you call in install_packages) should install the packer binary. 
I don't understand why the hashicorp repo needs to be added here  ? What am I missing ?

supritai replied:
Got your point. No need to hashicorp repo. Added condition to install additional packages for s390x only

bpradipt commented on config/peerpods/podvm/libvirt-podvm-image-handler.sh:
```diff
@@ -0,0 +1,280 @@
+    fi
+
+    # set a correspond qemu-system-* named link to qemu-kvm
+    ln -s /usr/libexec/qemu-kvm /usr/bin/qemu-system-$(uname -m)
```
You'll need to quote `"$(uname -m)` .. Applies to the entire code.

supritai replied:
ya, replaced with "$ARCH"

bpradipt commented on controllers/image_generator.go:
```diff
@@ -295,6 +301,11 @@ func (r *ImageGenerator) createJobFromFile(jobFileName string) (*batchv1.Job, er
 			})
+		} else if r.provider == LibvirtProvider {
+			job.Spec.Template.Spec.Containers[0].Env = append(job.Spec.Template.Spec.Containers[0].Env, corev1.EnvVar{
+				Name:  "LIBVIRT_IMAGE_ID",
```
s/LIBVIRT_IMAGE_ID/IMAGE_ID" based on previous comments w.r.to the var for delete job

supritai replied:
ok

bpradipt commented on controllers/image_generator.go:
```diff
@@ -575,6 +588,10 @@ func (r *ImageGenerator) validatePeerPodsConfigs() error {
+	// libvirt Secret Keys
+	libvirtSecretKeys := []string{"CLOUD_PROVIDER", "LIBVIRT_URI", "LIBVIRT_POOL", "LIBVIRT_VOL_NAME"}
+	// libvirt ConfigMap Keys
+	libvirtConfigMapKeys := []string{"CLOUD_PROVIDER", "PROXY_TIMEOUT"}
```
Do you need PROXY_TIMEOUT as mandatory key ?

supritai replied:
It is required for cloud-api-adaptor

bpradipt replied:
There is a default already set. You can try without this parameter in your env.

supritai replied:
okay, removing it

bpradipt commented on controllers/utils.go:
```diff
@@ -171,7 +171,10 @@ func getCloudProviderFromInfra(c client.Client) (string, error) {
+		return "", fmt.Errorf("Infrastructure.status.platformStatus is empty and is not libvirt")
+	} else if string(infrastructure.Status.PlatformStatus.Type) == "None" {
+		// For libvirt provider, PlatformStatus.Type is empty string, so returning libvirt provider
+		return "libvirt", nil
```
```suggestion
                return LibvirtProvider, nil
```

supritai replied:
ok

bpradipt commented:
@supritai once you have resolved a specific comment, can you please click on the associated "Resolve conversation" button, so that it's evident which comments have been addressed and what is still pending.

supritai committed:
Add libvirt provider support for image building

Added a new handler for libvirt to build the qcow2 image.
New ConfigMap added for libvirt specific data.
Modified the podvm create job to accomodate libvirt as well.
Added required changes for the image_generator controller.

Signed-off-by: Suprit-Iti <suprit.iti@ibm.com>

supritai commented:
> @supritai once you have resolved a specific comment, can you please click on the associated "Resolve conversation" button, so that it's evident which comments have been addressed and what is still pending.

Sure @bpradipt , most of them I have marked as resolved. Only where agreement of reviewer needed are not closed

bpradipt commented on config/peerpods/podvm/lib.sh:
```diff
@@ -110,9 +119,14 @@ install_binary_packages() {
-                mv "${install_dir}/yq_linux_amd64" "${install_dir}/yq"
+            # yq extracted file name is yq_linux_${ARCH}. Rename it
+            if [[ "${package_name}" == "yq" ]]; then
+                if [ "${ARCH}" == "s390x" ]; then
```
Can you please try with this one liner. 
`mv "${install_dir}"/yq_linux_"${ARCH/x86_64/amd64}" "${install_dir}"/yq`

supritai replied:
sure, thanks it reduced the code

bpradipt commented:
Thanks @supritai for addressing the comments. 
I added a minor suggestion to incorporate.  In case you have tried the latest code on s390x, can you post the test results.

bpradipt commented on config/peerpods/podvm/libvirt-podvm-image-handler.sh:
```diff
@@ -0,0 +1,280 @@
+    export GOPATH="/src"
+
+    if [ "${ARCH}" == "s390x" ]; then
+        # Install packer. Packer doesn't does not have prebuilt s390x arch binaries above Packer version 0.1.5 
```
```suggestion
        # Build packer from source for s390x as there is no prebuilt binary for the required packer version
```

supritai replied:
okay

bpradipt commented on config/peerpods/podvm/libvirt-podvm-image-handler.sh:
```diff
@@ -0,0 +1,280 @@
+	    sed -i -- "s/ALL_XC_OS=.*/ALL_XC_OS=\"Linux\"/g" scripts/build.sh
+        make bin && cp bin/packer /usr/local/bin/
+
+        # Install umoci source as binary for s390x is not available
```
```suggestion
        # Build umoci from source for s390x as there is no prebuilt binary
```

supritai replied:
okay

bpradipt commented on config/peerpods/podvm/libvirt-podvm-image-handler.sh:
```diff
@@ -0,0 +1,280 @@
+    # set a correspond qemu-system-* named link to qemu-kvm
+    ln -s /usr/libexec/qemu-kvm /usr/bin/qemu-system-"${ARCH}"
+
+    # cloud-utils package is not available for rhel.
```
```suggestion
    # Build cloud-utils package from source as prebuilt binary is not available
```

supritai replied:
sure

snir911 commented on config/peerpods/podvm/osc-podvm-delete-job.yaml:
```diff
@@ -27,6 +27,8 @@ spec:
             - name: IMAGE_ID
               value: "" # Set this to the azure image id to delete
+            - name: IMAGE_ID
+              value: "" # Set this to the libvirt image id to delete
```
Maybe LIBVIRT_VOL_NAME? because this is what will actually be deleted and to differ from Azure? (or LIBVIRT_IMAGE_ID to support future usage as you mentioned)

supritai replied:
Better to have LIBVIRT_IMAGE_ID now to avoid confusion of image and volume name.

snir911 commented on config/peerpods/podvm/libvirt-podvm-image-handler.sh:
```diff
@@ -0,0 +1,280 @@
+    # Add the libvirt image id to peer-pods-cm configmap
+    kubectl patch configmap peer-pods-cm -n openshift-sandboxed-containers-operator \
+    --type merge -p "{\"data\":{\"LIBVIRT_IMAGE_ID\":\"${LIBVIRT_VOL_NAME}\"}}" ||
+        error_exit "Failed to add the libvirt image id to peer-pods-cm configmap"
```
I think UPDATE_PEERPODS_CM is ignored at creation time, shouldn't it avoid updating CM if set to false?

supritai replied:
I added, thanks for pointing it

snir911 reviewed (commented):
added minor comments, overall LGTM

supritai reviewed (commented):
Attached the logs.
[osc-podvm-image-creation.log](https://github.com/user-attachments/files/16330389/osc-podvm-image-creation.log)
[osc-podvm-image-deletion.log](https://github.com/user-attachments/files/16330390/osc-podvm-image-deletion.log)

bpradipt reviewed (approved):
/lgtm
Thanks @supritai

snir911 reviewed (approved):
LGTM, thank you
